app_dart/lib/src/service/luci.dart

```diff
@@ -49,6 +49,42 @@ class LuciService {
   /// The AppEngine context to use for requests. Guaranteed to be non-null.
   final ClientContext clientContext;
 
+  /// Gets the list of recent LUCI tasks, broken out by the [BranchLuciBuilder]
+  ///  that owns them.
+  ///
+  /// The list of known LUCI builders is specified in [LuciBuilder.all].
+  Future<Map<BranchLuciBuilder, Map<String, List<LuciTask>>>>
+      getBranchRecentTasks({
+    String repo,
+    bool requireTaskName = false,
+  }) async {
```

Comment:
Nit: the return value of this function is too complicated. We can add `LuciBuilder` to `LuciTask` since every LUCI task has to run on a builder. Then this return value can be simplified to a list of LUCI tasks without two levels of nested mapping.

Reply:
I agree that will make the returning list simple, but it will make the follow-up status update challenging.

If you look into the update process, you will find that `luciBuilder` is the key base for looping around. Iterating through each `luciBuilder`, we then query DB for that builder. At the same time, `luciTask` needs to be ordered chronologically according to `luciBuilder`, and even a specific `sha`.

This is designed to reduce the time complexity.

Reply:
Exactly, the double nested mapping only makes sense in the high level refresh_chromebot_status, while this low level `getBranchRecentTasks` function is overfitting that. They are tightly coupled.

Reply:
Understand. Low level functions are not supposed to be used by a single target, the more general the better.

Per that, in the beginning, I tried using existing function `getRecentTasks` to implement the API, however, as @godofredoc pointed out, that resulted in too many loop cycles and complicated condition logics. It definitely will cause potential issues.

To solve that, here comes the specific function for our purpose here. Though tightly coupled with the high level purpose, it eases the logic greatly: reducing both loop cycles and logic conditions.

Reply:
There's no extra loop cycles, it's just a choice of having them in the low level API or high level functions. The current approach is less readable as one has to read both functions to understand the structure of the return value, and the reason why it's structured that way.

Reply:
There's no extra loop cycles, it's just a choice of having them in the low level API or high level functions. The current approach is less readable as one has to read both functions to understand the structure of the return value, and the reason why it's structured that way.